File: server/lib/jiosaavn.js
import axios from "axios";
import { JIO_SAAVN_PLAYLIST_SEARCH, JIO_SAAVN_SONG_SEARCH } from "../utils/constant.js";
import { checkSimilarity } from "../utils/utils.js";
import logger from "../utils/logger.js";

class JioSaavn {
    async getSongBySongName(songName, retryCount = 1) {
        try {
            const response = await axios.get(JIO_SAAVN_SONG_SEARCH(songName));
            const results = response.data.results.find(track => checkSimilarity(songName, track.title) > 60);
            if (!results) return;
            const moreInfo = results.more_info;

            if(moreInfo.duration > 600){
                throw new Error("Song Duration is more than 10 minutes.")
            }
            return {
                title: results.title,
                url: moreInfo.encrypted_media_url,
                duration: moreInfo.duration
            }
        } catch (error) {
            logger.error(error);
            logger.error("Failed after retrying:", { error });
            return;
        }
    }

    async getPlaylistDetail(playlistId) {
        try {
            const response = await axios.get(JIO_SAAVN_PLAYLIST_SEARCH(playlistId));
            const { list } = response.data;
            if(list.length <= 0){
                throw new Error("Invalid Playlist ID");
            }
            return list;
        } catch (error) {
            logger.error(error);
            logger.error("Error fetching Playlist");    
            return null;
        }
    }
}

export default JioSaavn;

File: app/integrations/jiosaavn.py
import aiohttp
from app.core import logger
from app.core.utils import check_similarity
from app.core.constants import JIO_SAAVN_SONG_SEARCH, JIO_SAAVN_PLAYLIST_SEARCH

class JioSaavn:
    async def get_song_by_song_name(self, song_name: str, retry_count: int = 1):
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(JIO_SAAVN_SONG_SEARCH(song_name)) as response:
                    data = await response.json()
                    
                    if not data.get('results'):
                        return None
                    
                    results = None
                    for track in data['results']:
                        if check_similarity(song_name, track.get('title', '')) > 60:
                            results = track
                            break
                    
                    if not results:
                        return None
                    
                    more_info = results.get('more_info', {})
                    
                    if more_info.get('duration', 0) > 600:
                        raise Exception("Song Duration is more than 10 minutes.")
                    
                    return {
                        'title': results.get('title'),
                        'url': more_info.get('encrypted_media_url'),
                        'duration': more_info.get('duration')
                    }
        except Exception as error:
            logger.error(str(error))
            logger.error("Failed after retrying", error=str(error))
            return None
    
    async def get_playlist_detail(self, playlist_id: str):
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(JIO_SAAVN_PLAYLIST_SEARCH(playlist_id)) as response:
                    data = await response.json()
                    
                    playlist_list = data.get('list', [])
                    if len(playlist_list) <= 0:
                        raise Exception("Invalid Playlist ID")
                    
                    return playlist_list
        except Exception as error:
            logger.error(str(error))
            logger.error("Error fetching Playlist")
            return None
